src/reflect/scala/reflect/internal/util/Parallel.scala

```diff
@@ -0,0 +1,101 @@
+package scala.reflect.internal.util
+
+import java.util.concurrent.atomic.AtomicInteger
+
+object Parallel {
+
+  class Counter {
+    private val count = new AtomicInteger
+
+    def get: Int = count.get()
+
+    def reset(): Unit = {
+      assertOnMain()
+      count.set(0)
+    }
+
+    def incrementAndGet(): Int = count.incrementAndGet
+
+    def getAndIncrement(): Int = count.getAndIncrement
+
+    override def toString: String = s"Counter[$count]"
+  }
+
+  // Wrapper for `synchronized` method. In future could provide additional logging, safety checks, etc.
+  def synchronizeAccess[T <: Object, U](obj: T)(block: => U): U = {
+    obj.synchronized[U](block)
+  }
+
+  def WorkerThreadLocal[T <: AnyRef](valueOnWorker: => T, valueOnMain: => T) = new WorkerOrMainThreadLocal[T](valueOnWorker, valueOnMain)
+
+  def WorkerThreadLocal[T <: AnyRef](valueOnWorker: => T) = new WorkerThreadLocal[T](valueOnWorker)
+
+  // `WorkerOrMainThreadLocal` allows us to have different type of values on main and worker threads.
+  // It's useful in cases like reporter, when on workers we want to just store messages and on main we want to print them,
+  class WorkerOrMainThreadLocal[T](valueOnWorker: => T, valueOnMain: => T) {
+
+    private var main: T = null.asInstanceOf[T]
+
+    private val worker: ThreadLocal[T] = new ThreadLocal[T] {
+      override def initialValue(): T = valueOnWorker
+    }
+
+    final def get: T = {
+      if (isWorker.get()) worker.get()
+      else {
+        if (main == null) main = valueOnMain
+        main
+      }
+    }
+
+    final def set(value: T): Unit = if (isWorker.get()) worker.set(value) else main = value
+
+    final def reset(): Unit = {
+      worker.remove()
+      main = valueOnMain
+    }
+  }
+
+  // `WorkerThreadLocal` detects reads/writes of given value on the main thread and
+  // and report such violations by throwing exception.
+  class WorkerThreadLocal[T](valueOnWorker: => T)
+    extends WorkerOrMainThreadLocal(valueOnWorker, throw new IllegalStateException("not allowed on main thread"))
+
+  // Asserts that current execution happens on the main thread
+  def assertOnMain(): Unit = {
+    assert(!isWorker.get())
+  }
+
+  // Asserts that current execution happens on the worker thread
+  def assertOnWorker(): Unit = {
+    assert(isWorker.get())
+  }
+
+  // Runs block of the code in the 'worker thread' mode
+  // All unit processing should always happen in the worker thread
+  @inline final def asWorkerThread[T](fn: => T): T = {
+    val previous = isWorker.get()
```

Comment:
do we need this to be reenterent?

Reply:
It's not so much about being re-entrant as for cases when we have:
1) main execution
2) worker execution
3) main execution
happening on the same thread (tests, phases with disabled parallel compilation)